### ISSUE TITLE
Spurious `Warning: Unnamed return variable can remain unassigned.` produced by `while (true) { ... return; }`

### ISSUE THREAD

**Issue by duncancmt** [ethereum/solidity#13746]
## Description

The warning `Warning: Unnamed return variable can remain unassigned.` is produced by code where it is categorically impossible for the return variable to remain unassigned.

## Environment

- Compiler version: 0.8.17
- Target EVM version (as per compiler settings): London
- Framework/IDE (e.g. Truffle or Remix): Foundry
- EVM execution environment / backend / blockchain client: N/A
- Operating system: Fedora GNU/Linux

## Steps to Reproduce

```solidity
pragma solidity 0.8.17;

contract Repro {
  function repro(uint256 x) external pure returns (bool) {
    while (true) {
      if ((x >>= 1) == 0) {
        return true;
      }
    }
  }
}
```

```
Compiler run successful (with warnings)
warning[6321]: Warning: Unnamed return variable can remain unassigned. Add an explicit return with value to all non-reverting code paths or name the variable.
 --> contracts/Repro.sol:4:52:
  |
4 |   function repro(uint256 x) external view returns (bool) {
  |                                                    ^^^^
```

**Comment by cameel** [comment]
Simpler repro:
```solidity
function f() pure returns (bool) {
    while (true)
        return true;
}
```
This warning has been there since 0.7.4. There's no warning on 0.7.3.

It's not really a bug though - more of a limitation of the current control flow graph implementation. The CFG does not evaluate conditions so it can't tell whether the code after the loop will ever be reached. Conservatively, it has to assume that it's possible. It's the same with things like `if (true)`.

As a workaround, you could help the compiler figure this out by making it clear that the function would never return the value even if the loop ended:
```solidity
function f() pure returns (bool) {
    while (true)
        return true;
    revert();
}
```

Unfortunately it impossible to evaluate conditions in a general case. We could do that for some limited situations, like constant expressions, but it's debatable if the gains here are worth the effort. We use the CFG in the optimizer so theoretically it would let us detect more code as unreachable and remove it, but, realistically, I doubt that leaving unreachable code after an infinite loop is that common other than by a mistake (which can happen but is probably not common). Unless #13724 is going to make this actually simple to implement?

One thing that I think would actually be helpful would be to evaluate specifically `true` and `false`. This would provide a more obvious workaround (i.e. if you know the expression in the loop condition is always true, you can specify just that). In fact, I originally wanted to use `assert(false)` in my workaround, which would be more semantic and that change would help here. What do you think @ekpyron?

**Comment by ekpyron** [comment]
I mean, another quick and dirty workaround for the warning here would be to give the return variable a name :-).

But yeah, I agree with what @cameel says. In general there is no way to be completely correct about this (you can construct cases in which, while you can prove that something is always fine, that proof can get arbitrarily complex).

@cameel However, the CFG used for analysis is not the CFG we use for optimization (which happens on the Yul level). The analysis in the optimizer should in fact be powerful enough to catch dead code after infinite loops and remove it. Which should make the ``revert();`` workaround zero-cost in practice.

In any case, we won't start particular special casing on verbatim expressions (like ``true``) here, since that's generally a bottomless pit. We can indeed re-evaluate this once #13724 is done, but until then, I'm unfortunately closing this as won't-fix.

**Comment by gouldra** [comment]
I solved this problem using compiler version 8.17.
Thanks
